### DISCUSSION TITLE
when will glfw be included in the binary builds?

### DISCUSSION THREAD

**Discussion by odarcan** [crazywhalecc/static-php-cli#191] [Q&A] [answered]
Hi I assume your builds dont have php-glfw in them yet, I tried to run some code that uses that extension, I got some errors.
When will they be included?

thanks!

hmm it says it has glfw in it but cannot be linked statically? what does that mean

I have the extension added to my regular php and it works, Id like to get yours working too

@crazywhalecc

**Comment by crazywhalecc** [comment]
glfw depends on OpenGL(required) and X11(alternative in Linux). These libraries cannot be compiled statically, or it's very difficult. On macOS platform, we can dynamically link the system framework and ship it with no pain, just not use the old version of macOS.

**Reply by odarcan** [reply]
@crazywhalecc 
I understood some words there, I am using mint myself

what do you need, a team of experts looking into this?
I can ask around if I knew the question.

**Reply by odarcan** [reply]
Would 10 $ solve the problem?

**Reply by crazywhalecc** [reply]
Actually, it is not difficult to statically compile glfw to php. The difficulty is that it is difficult to statically compile libraries such as OpenGL that involve GPU and system hardware interfaces. So if you really want to use these system-level libraries, it is best to compile them dynamically.

However, for Linux, we can also achieve partial dynamic linking of some libraries that cannot be statically compiled without completely statically compiling PHP, just like macOS. But there are certain requirements for doing so: for example, the system must install a specified version of the OpenGL library in advance. Compared with compiling a dynamically linked PHP directly in the local environment, it may not be worth it.

**Reply by crazywhalecc** [reply]
I mentioned a similar idea in #58, which is to compile with glibc-gcc under Linux and statically link all other libraries except glibc. The advantage of compiling in this way is that you can distribute and use it under a system that uses gnu-libc as libc, and supports dynamic loading of external extensions. But this also imposes requirements on the system environment, and is of little help to extensions such as glfw that use system libraries.

**Reply by odarcan** [reply]
There has to be a way,
I have a php installed with glfw, and it starts 3d programs written in php, instantly and I think it has potential to be a new cross-os game platform thing, the php code of the program could be written once and distributed to many operating systems,since you have already made it for mac and win, what remains is linux, then I can keep nagging for whatever else is missing.

bundling the php code with different executables for different operating systems could be automated on a website for example

@crazywhalecc

**Reply by odarcan** [reply]
Can you write down the exact problem with details, so I can share it around with people who may come up with a solution? 

Is your problem that your solution imposes requirements on the system environment?

Then can you for example make one that works for me for example?
Can you make one that works for latest Mint users?
or one for latest Ubuntu users?

**Reply by odarcan** [reply]
@crazywhalecc 

also: so this is a linux / opengl problem, correct? not a php problem ?

**Reply by crazywhalecc** [reply]
Exactly. php-glfw extension and php itself can be easily statically compiled. Many of the extensions I completed before that took a lot of time to implement were also due to this: static compilation of dependent libraries is troublesome.

**Comment by jingjingxyk** [comment]
linux build reference :   Off-screen Rendering.  

https://docs.mesa3d.org/osmesa.html
https://stackoverflow.com/questions/10209663/statically-linked-opengl-library-on-windows

**Comment by odarcan** [comment]
thanks,
If "mesa3d" was used, would it work?
@jingjingxyk  @crazywhalecc

**Reply by crazywhalecc** [reply]
Compiling mesa3d is another big problem. It uses meson for building, which I'm not very familiar with. 

For now I have a question: Do you need just a standalone php file or a php binary distribution that is portable to almost any Linux?

**Reply by odarcan** [reply]
With glfw I found that writing 3d games can be incredibly easy,
I would like to be able to distribute it to others without having them install the glfw library or php (if they dont have it already), also there is the problem of having correct version of php, these would all hinder easy distribution.

For starters, I would like a binary that can run on linux mint, my distro. 
but main thing is the glfw part, can it be included for mint?




------------------------------

I also have an idea for making apps in php, in Android.
Can you make binaries for android?
if yes I will explain more.

**Reply by odarcan** [reply]
Do you have an email address?

**Reply by crazywhalecc** [reply]
I understand. It doesn't sound like a scenario that completely requires static compilation. What is more needed is a system-level PHP migration solution. This idea I mentioned in #58 is basically to build a standalone PHP binary that can rely on some system libraries.

But unfortunately, even so, trying to statically link these huge system-level underlying UI libraries is tantamount to packaging half of the operating system into a PHP binary, which is very unrealistic. If you want to implement transplantation to different platforms or distributions, there is an idea to consider: making a PHP distribution just like making `.deb` software. When you publish, you can write automatic build processes for multiple distributions and use them by different people. You need to download different versions of php files.

In short:

If your main consideration is binary portability, all cases that rely on static compilation are the most convenient for portability, but some libraries (such as X11, mesa3d, OpenGL, etc. that glfw depends on) are almost difficult to compile statically. If you are only running under a specific distribution (for example, only Ubuntu with the Xorg desktop environment), then you can compile it under Ubuntu in the same environment, and then distribute the php binary with glfw and let it use the system's dynamic library.

**Comment by odarcan** [comment]
@crazywhalecc 
https://www.glfw.org/

**Comment by crazywhalecc** [comment]
Looks like almost-statically linked binaries for Linux and macOS is possible to add glfw in the php-cli or micro.sfx. But yet I haven't tried it.

Alternatively, the latest static-php-cli can build shared extension too (portable so file), that may also help glfw extension build for Linux.

**Comment by henderkes** [comment, marked as answer]
I took a look at this and it has an insane dependency tree that, among others, also includes the whole massive X11 system. The only feasible solution for this is to link against systen glfw dynamically, which can be done with SPC_TARGET="musl -dynamic" for Alpine or SPC_LIBC=glibc for glibc systems like Debian, Ubuntu, Mint, RHEL or anything in between. You need to add glfw ext to the project yourself and use PKG_CONFIG_PATH=/usr/lib(64)/pkgconfig bin/spc build ...

But you do not end up with a portable binary. Every user will then need to have the correct glfw version installed to even run your php binary for anything.

**Comment by crazywhalecc** [comment]
Looks like we made it, portable glfw app on Linux!

<img width="1426" height="1332" alt="image" src="https://github.com/user-attachments/assets/30b0543a-9f90-435d-a517-75bcc7b4b495" />

**Reply by crazywhalecc** [reply]
But anyway it needs some patches and specific build flags and need to build with glibc. Confirmed that it can run on another Linux OS with desktop installed.

**Comment by crazywhalecc** [comment]
All platforms are supported on v3.

**Comment by odarcan** [comment]
I am trying this now, it says call to undefined function glfwInit() 
I downloaded php-8.5.5-cli-linux-x85_64
which doesnt have it yes?

i tried your  https://static-php.dev/en/guide/cli-generator.html
there is no glfw yet yes?

I manually added glfw
final message was
library [glfw] is in the lib.json list but not supported to build on Linux.

I am not sure how this works, care to explain?
how do I get to that purple screen above?
could you post the binary for convenience?

I am on linux mint.

**Reply by crazywhalecc** [reply]
This feature is only available on v3 version, which is not released officially yet. We are going to publish alpha1 version recently, just stay tuned.

Or you can try branch `v3-dev`, and uses the same command line. Before building, edit `env.ini` `[linux]` section, make `SPC_TARGET=native-native-gnu.2.17`. Before you build, just confirm that your OS has installed X11 first.

**Reply by henderkes** [reply]
X11-dev, X11 alone is only enough during runtime, not for building.
